web/packages/teleport/src/DesktopSession/KeyboardHandler.tsx

```diff
@@ -0,0 +1,271 @@
+/**
+ * Teleport
+ * Copyright (C) 2023  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { Platform, getPlatform } from 'design/platform';
+
+import { TdpClient, ButtonState } from 'teleport/lib/tdp';
+import { SyncKeys } from 'teleport/lib/tdp/codec';
+
+/**
+ * Handles keyboard events.
+ */
+export class KeyboardHandler {
+  /**
+   * Cache for keys whose down event has been withheld until the next keydown or keyup,
+   * or to never be sent if this cache is cleared onfocusout.
+   */
+  private withheldDown: Map<string, boolean> = new Map();
+  /**
+   * Cache for keys whose up event has been delayed for 10ms.
+   */
+  private delayedUp: Map<string, NodeJS.Timeout> = new Map();
+  /**
+   * Tracks whether the next keydown or keyup event should sync the
+   * local toggle key state to the remote machine.
+   *
+   * Set to true:
+   * - On component initialization, so keys are synced before the first keydown/keyup event.
+   * - On focusout, so keys are synced when the user returns to the window.
+   */
+  private syncBeforeNextKey: boolean = true;
+  private isMac: boolean = getPlatform() === Platform.macOS;
+
+  /**
+   * Primary method for handling keyboard events.
+   */
+  public handleKeyboardEvent(
```

Comment:
You did a really nice job breaking this class up into small and easily comprehensible functions.

Seems like a great candidate for some unit tests, no?